File: Code/Asteroid_Trajectory_Selection/GREEDY/greedy_selector.m


try
    asteroid_list = LOAD_KERNELS("NOTABLE_ASTEROID_BSPs");
catch exception
    close all force;
    fprintf('An error occurred: %s\n', exception.message);
    error("Most likely you need to add NOTABLE_ASTEROID_BSPs to your path" + ...
            " and you must be in the main github folder.");
end

M_VALUES = [0.1, -0.1, 1, -1];
M_length = length(M_VALUES);

MIN_LAUNCH_DATE_UTC = 'Jan 1 12:00:00 UTC 2027';
MAX_LAUNCH_DATE_UTC = 'Dec 31 12:00:00 UTC 2035';

iter_vec = [1, 1, 1, 1, 1, 1];

while iter_vec(end) <= M_length
    

    M = reshape(M_VALUES(iter_vec), 2, 3)';

    save_filename = "greedy_test";

    for i = 1:6
        save_filename = save_filename + sprintf("_%.1f", M_VALUES(iter_vec(i)));
    end

    asteroid_optimized_data = GENERATE_GREEDY_OPTIMIZED_DATA(asteroid_list, ...
        M, MIN_LAUNCH_DATE_UTC, MAX_LAUNCH_DATE_UTC, save_filename);

    iter_vec(1) = iter_vec(1) + 1;
    while any(iter_vec > M_length)
        for j = 1:(length(iter_vec)-1)
            if iter_vec(j) > M_length
                iter_vec(j+1) = iter_vec(j+1) + 1;
                iter_vec(j) = 1;
            end
        end
        if iter_vec(end) > M_length
            break
        end
    end
end

File: Python/asteroid_trajectory_selection/greedy/greedy_selector.py
"""Iterate over all M_VALUES combinations and run greedy optimization."""

import itertools
import numpy as np

from ..load_kernels import load_kernels
from .generate_greedy_optimized_data import generate_greedy_optimized_data


def main():
    """Main entry point for the greedy selector script."""
    try:
        asteroid_list = load_kernels("NOTABLE_ASTEROID_BSPs", "generic_kernels")
    except Exception as e:
        print(f'An error occurred: {e}')
        raise RuntimeError(
            "Most likely you need to add NOTABLE_ASTEROID_BSPs to your path"
            " and you must be in the main github folder.")

    M_VALUES = [0.1, -0.1, 1, -1]
    M_length = len(M_VALUES)

    MIN_LAUNCH_DATE_UTC = 'Jan 1 12:00:00 UTC 2027'
    MAX_LAUNCH_DATE_UTC = 'Dec 31 12:00:00 UTC 2035'

    # Iterate over all combinations of 6 indices into M_VALUES
    # This replicates the MATLAB odometer-style iteration
    iter_vec = [0, 0, 0, 0, 0, 0]  # 0-indexed

    while iter_vec[5] < M_length:
        # Build M matrix: reshape 6 values into (3, 2)
        m_vals = [M_VALUES[idx] for idx in iter_vec]
        M = np.array(m_vals).reshape(3, 2)

        save_filename = "greedy_test"
        for idx in iter_vec:
            save_filename += f"_{M_VALUES[idx]:.1f}"

        generate_greedy_optimized_data(
            asteroid_list, M,
            MIN_LAUNCH_DATE_UTC, MAX_LAUNCH_DATE_UTC,
            save_filename)

        # Increment odometer (carry from left to right)
        iter_vec[0] += 1
        for j in range(len(iter_vec) - 1):
            if iter_vec[j] >= M_length:
                iter_vec[j + 1] += 1
                iter_vec[j] = 0
        if iter_vec[-1] >= M_length:
            break


if __name__ == '__main__':
    main()
